lib/spack/spack/database.py

```diff
@@ -913,6 +921,11 @@ def _write(self, type, value, traceback):
             with open(temp_file, 'w') as f:
                 self._write_to_file(f)
             os.rename(temp_file, self._index_path)
+            if _use_uuid:
+                with open(self._verifier_path, 'w') as f:
+                    new_verifier = str(uuid.uuid4())
+                    f.write(new_verifier)
+                    self.last_seen_verifier = new_verifier
```

Comment:
@adamjstewart @tgamblin @aweits Apparently when we reach this line we don't update `self._data` and we have therefore an inconsistent view of the DB later on (since some specs might be referencing roots that have been removed). Removing this line seems to fix #15773. I'll check  if we can come up with a more performant update of `self._data` than re-reading from file.